### PR TITLE
Log errors in handler of health checks.

### PR DESCRIPTION
**Pull request by amikhailau**

### CODE REVIEW

File: health/handler.go [hunk 1]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
 func (ch *checksHandler) AddLiveness(name string, check Check) {
 	ch.lock.Lock()
 	defer ch.lock.Unlock()
+	ch.logger.WithFields(logrus.Fields{
+		"Liveness path":  ch.livenessPath,
+		"Readiness path": ch.readinessPath,
+		"Name":           name,
+	}).Info("adding liveness check")
```
Comment (wplu) [inline health/handler.go]: i would prefer this to be warn so they show up in production

File: health/handler.go [hunk 2]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
 }
 
 func (ch *checksHandler) AddReadiness(name string, check Check) {
 	ch.lock.Lock()
 	defer ch.lock.Unlock()
+	ch.logger.WithFields(logrus.Fields{
+		"Liveness path":  ch.livenessPath,
+		"Readiness path": ch.readinessPath,
+		"Name":           name,
+	}).Info("adding readiness check")
```
Comment (wplu) [inline health/handler.go]: i would prefer this to be warn so they show up in production

File: health/handler.go [hunk 3]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
 
 	ch.readinessChecks[name] = check
 }
 
 // Handler returns a new http.Handler for the given health checker
 func (ch *checksHandler) Handler() http.Handler {
+	ch.logger.WithFields(logrus.Fields{
+		"Liveness path":  ch.livenessPath,
+		"Readiness path": ch.readinessPath,
+	}).Info("creating new ServeMux for health checker")
```
Comment (wplu) [inline health/handler.go]: i would prefer this to be warn so they show up in production

File: health/handler.go [hunk 4]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
 	ch.registerMux(mux)
 	return mux
 }
 
 // RegisterHandler registers the given health and readiness patterns onto the given http.ServeMux
 func (ch *checksHandler) RegisterHandler(mux *http.ServeMux) {
+	ch.logger.WithFields(logrus.Fields{
+		"Liveness path":  ch.livenessPath,
+		"Readiness path": ch.readinessPath,
+	}).Info("registering ServeMux for health checker")
```
Comment (wplu) [inline health/handler.go]: i would prefer this to be warn so they show up in production

File: health/handler.go [hunk 5]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
 	ch.registerMux(mux)
 }
 
 func (ch *checksHandler) registerMux(mux *http.ServeMux) {
+	ch.logger.WithFields(logrus.Fields{
+		"Liveness path":  ch.livenessPath,
+		"Readiness path": ch.readinessPath,
+	}).Info("registering endpoints for health checker")
```
Comment (wplu) [inline health/handler.go]: i would prefer this to be warn so they show up in production

File: health/handler.go [hunk 6]
```diff
@@ -83,6 +110,10 @@ func (ch *checksHandler) readyEndpoint(rw http.ResponseWriter, r *http.Request)
 
 func (ch *checksHandler) handle(rw http.ResponseWriter, r *http.Request, checksSets ...map[string]Check) {
 	if r.Method != http.MethodGet {
+		ch.logger.WithFields(logrus.Fields{
+			"Liveness path":  ch.livenessPath,
+			"Readiness path": ch.readinessPath,
+		}).Info("received non-GET request")
```
Comment (wplu) [inline health/handler.go]: this should be Error instead of Info

Comment (wplu) [reply health/handler.go]: please also log the r.Path

File: health/handler.go [hunk 7]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
 func (ch *checksHandler) AddLiveness(name string, check Check) {
 	ch.lock.Lock()
 	defer ch.lock.Unlock()
+	ch.logger.WithFields(logrus.Fields{
+		"Liveness path":  ch.livenessPath,
```
Comment (wplu) [inline health/handler.go]: change `Liveness path` to snake case `liveness_path`

File: health/handler.go [hunk 8]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
+		"Readiness path": ch.readinessPath,
```
Comment (wplu) [inline health/handler.go]: change "Readiness path" to snake_case

File: health/handler.go [hunk 9]
```diff
@@ -98,6 +129,10 @@ func (ch *checksHandler) handle(rw http.ResponseWriter, r *http.Request, checksS
 				continue
 			}
 			if err := check(); err != nil {
+				ch.logger.WithFields(logrus.Fields{
+					"Liveness path":  ch.livenessPath,
+					"Readiness path": ch.readinessPath,
+				}).WithError(err).Error("health check returned error")
```
Comment (wplu) [inline health/handler.go]: add "path": r.Path

File: health/handler.go [hunk 10]
```diff
@@ -98,6 +129,10 @@ func (ch *checksHandler) handle(rw http.ResponseWriter, r *http.Request, checksS
 				continue
 			}
 			if err := check(); err != nil {
+				ch.logger.WithFields(logrus.Fields{
```
Comment (wplu) [inline health/handler.go]: add "path": r.Path as another field

File: health/handler.go (hunk 1 above)
Comment (drewwells) [reply health/handler.go]: It's not a pseudo error though, isn't `info` the default log level?

File: health/handler.go (hunk 6 above)
Comment (drewwells) [reply health/handler.go]: `warn` in my opinion

Comment (drewwells) [reply health/handler.go]: `warn` in my opinion

File: health/handler.go [hunk 11]
```diff
@@ -83,6 +110,11 @@ func (ch *checksHandler) readyEndpoint(rw http.ResponseWriter, r *http.Request)
 
 func (ch *checksHandler) handle(rw http.ResponseWriter, r *http.Request, checksSets ...map[string]Check) {
```
Comment (drewwells) [inline health/handler.go]:
```suggestion
func checkLogger(logger *logrus.Entry) (*logrus.Entry, bool) {
	if logger == nil {
		return logger, false
	} else if logger.Logger.Out == ioutil.Discard {
		return logger, false
	}
	return logger, true
}

func (ch *checksHandler) handle(rw http.ResponseWriter, r *http.Request, checksSets ...map[string]Check) {
	logger := ch.logger
	ctxLogger, ok := checkLogger(ctxlogrus.Extract(r.Context()))
	if ok {
		logger = ctxLogger
	}
```

File: health/handler.go [hunk 12]
```diff
@@ -25,7 +29,7 @@ type Checker interface {
 
 // NewChecksHandler accepts two strings: health and ready paths.
 // These paths will be used for liveness and readiness checks.
-func NewChecksHandler(healthPath, readyPath string) Checker {
+func NewChecksHandler(healthPath, readyPath string, logger *logrus.Logger) Checker {
```
[the remark this answers is not in the document]
Comment (amikhailau) [reply health/handler.go]: This changes will be kept in a separate branch from master only for identity to use.

File: health/handler.go [hunk 13]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
 func (ch *checksHandler) AddLiveness(name string, check Check) {
 	ch.lock.Lock()
 	defer ch.lock.Unlock()
+	ch.logger.WithFields(logrus.Fields{
```
Comment (wplu) [inline health/handler.go]: surround this with if nil check

```
if ch.logger != nil {
  ch.logger...
}
```

File: health/handler.go [hunk 14]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
+		"Name":           name,
+	}).Warn("adding liveness check")
 
 	ch.livenessChecks[name] = check
 }
 
 func (ch *checksHandler) AddReadiness(name string, check Check) {
 	ch.lock.Lock()
 	defer ch.lock.Unlock()
+	ch.logger.WithFields(logrus.Fields{
```
Comment (wplu) [inline health/handler.go]: surround with nil check

File: health/handler.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (wplu) [reply health/handler.go]: do not change this function.

create new function using functional options:
```
func NewChecksHandlerWithOptions(healthPath, readyPath string, options ...func(Checker)) Checker {
```

File: health/handler.go [hunk 15]
```diff
@@ -37,6 +41,7 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
 		livenessChecks:  map[string]Check{},
 		readinessPath:   readyPath,
 		readinessChecks: map[string]Check{},
+		logger:          logger,
```
[the remark this answers is not in the document]
Comment (wplu) [reply health/handler.go]: default is nil - logger is only added if WithLogger(logger) functional option is called

File: health/handler.go [hunk 16]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
+		"readiness_path": ch.readinessPath,
+		"Name":           name,
+	}).Warn("adding readiness check")
 
 	ch.readinessChecks[name] = check
 }
 
 // Handler returns a new http.Handler for the given health checker
 func (ch *checksHandler) Handler() http.Handler {
+	ch.logger.WithFields(logrus.Fields{
```
Comment (wplu) [inline health/handler.go]: surround with nil check

File: health/handler.go [hunk 17]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
+		"readiness_path": ch.readinessPath,
+	}).Warn("creating new ServeMux for health checker")
 	mux := http.NewServeMux()
 	ch.registerMux(mux)
 	return mux
 }
 
 // RegisterHandler registers the given health and readiness patterns onto the given http.ServeMux
 func (ch *checksHandler) RegisterHandler(mux *http.ServeMux) {
+	ch.logger.WithFields(logrus.Fields{
```
Comment (wplu) [inline health/handler.go]: surround with nil check

File: health/handler.go [hunk 18]
```diff
@@ -45,30 +50,52 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
...
+		"liveness_path":  ch.livenessPath,
+		"readiness_path": ch.readinessPath,
+	}).Warn("registering ServeMux for health checker")
 	ch.registerMux(mux)
 }
 
 func (ch *checksHandler) registerMux(mux *http.ServeMux) {
+	ch.logger.WithFields(logrus.Fields{
```
Comment (wplu) [inline health/handler.go]: surround with nil check

File: health/handler.go [hunk 19]
```diff
@@ -83,6 +110,11 @@ func (ch *checksHandler) readyEndpoint(rw http.ResponseWriter, r *http.Request)
 
 func (ch *checksHandler) handle(rw http.ResponseWriter, r *http.Request, checksSets ...map[string]Check) {
 	if r.Method != http.MethodGet {
+		ch.logger.WithFields(logrus.Fields{
```
Comment (wplu) [inline health/handler.go]: surround with nil check

File: health/handler.go [hunk 20]
```diff
@@ -13,12 +17,15 @@ type checksHandler struct {
...
 	readinessChecks map[string]Check
+
+	logger *logrus.Logger
 }
 
 // Checker ...
 type Checker interface {
 	AddLiveness(name string, check Check)
 	AddReadiness(name string, check Check)
+	addLogger(*logrus.Logger)
```
Comment (wplu) [inline health/handler.go]: let's not do this

File: health/handler.go [hunk 21]
```diff
@@ -37,38 +44,91 @@ func NewChecksHandler(healthPath, readyPath string) Checker {
 		livenessChecks:  map[string]Check{},
 		readinessPath:   readyPath,
 		readinessChecks: map[string]Check{},
+		logger:          nil,
+	}
+
+	return ch
+}
+
+func NewChecksHandlerWithOptions(healthPath, readyPath string, options ...func(Checker)) Checker {
```
Comment (wplu) [inline health/handler.go]: func NewChecksHandlerWithOptions(healthPath, readyPath string, options ...func(* ChecksHandler)) *ChecksHandler {

File: health/handler.go (hunk 1 above)
Comment (wplu) [reply health/handler.go]: warn is the log level in production

File: health/handler.go [hunk 22]
```diff
@@ -26,63 +32,137 @@ type Checker interface {
...
+}
+
+func (ch *ChecksHandler) AddLiveness(name string, check Check) {
 	ch.lock.Lock()
 	defer ch.lock.Unlock()
+	if ch.logger != nil {
+		ch.logger.WithFields(logrus.Fields{
+			"liveness_path":  ch.livenessPath,
+			"readiness_path": ch.readinessPath,
+			"Name":           name,
```
Comment (drewwells) [inline health/handler.go]: lowercase `Name` please

**Pull request merged**